This past summer, I got the opportunity to be a course assistant at the Biology of Aging course 
at the Marine Biological Laboratory in Woods Hole. It was a fascinating course focused on aging 
research, that involved a variety of faculty who approached the problem of aging from multiple 
perspectives using different model organisms. One of the coolest approaches, in my opinion, was 
taken by Matt Harris and Stephen Treaster. They were doing comparative genomic analysis of 
different animals that demonstrated extreme longevity. It was a crazy cool approach, and 
although I didn’t understand a ton of what they did, and how they did it, I was fascinated. 
Additionally, I’m in a research lab here at W&M. I’m currently working on a bulk RNA-seq 
project, and I’m about to start the data analysis portion of the project. These experiences 
have given me an appreciation for the power and usefulness of bioinformatics, so when I saw 
this class, I had to take it. Bioinformatics is such a useful tool, and I want to get better at 
it. I’m currently planning on doing a PhD in biology and having the toolkit to do basic 
bioinformatics would be quite helpful in that pursuit. During the course I hope to develop a 
solid foundation of basic skills and knowledge in bioinformatics. I want to be able to both 
better understand bionformatic based research and to approach and tackle new bioinformatics 
projects of my own in the future. In pursuit of this I’m going to be very systematic in my 
approach to the assignments and the lessons. I’m going to try and review material often and 
become quite familiar with the approaches and information provided in class.

The folder structure I created lends itself well to reproducible research because it’s clear 
and well organized. It has a folder dedicated to data, allowing us to call the data if and when 
we need it, and allowing us to treat it as read only, preventing the possiblity of incorrectly 
editing the data itself. It has folders for docs, logs, config, scripts, and results. This 
allows for extensive documentation and provides a clear organiztional structure for where 
different pieces of information will be. This would make it easier for someone in the future to 
understand how the project was carried out and hopefully enable them to reproduce the findings 
of the project. Reproducibility also requires well documented code. Code in general can be 
difficult to read and understand. Documentation will ideally make the process of understanding 
the code much easier. This is helpful in two ways. First it makes the initial process of 
building out the project easier. If and when a person gets confused or gets a result that seems 
incorrect the researcher can fall back on their own documentation when reviewing their code. 
This can make the process of troubleshooting much easier. Second it makes it much easier for 
others to reproduce the work. In order to successfully reproduce a bioinformatics project 
whoever’s trying to reproduce it needs to understand how and why the code works and then 
successfully execute that code. Clear documentation makes this task much easier and allows the 
new researcher to more easily identify any points of trouble they may come across. 
